teps/0094-configuring-resources-at-runtime.md

```diff
@@ -52,12 +61,495 @@ While this may be a valuable feature, it should be considered in a separate prop
 - Kubeflow pipelines and other data pipelines may have variable resource requirements
 depending on the data being processed.
-- Catalog Tasks should be generally reusable in different environments
+- Catalog `Task`s should be generally reusable in different environments
 that may have different resource constraints.
 
 ## Requirements
 
 - Users can specify `Step` and `Sidecar` resource requirements at runtime.
 - Users can specify `Step` and `Sidecar` resource requirements for `Task`s 
 or `Pipeline`s they don't own, especially those in the Catalog.
-- Users can specify resource requirements for individual `Step`s and `Sidecar`s.
\ No newline at end of file
+- Users can specify resource requirements for individual `Step`s and `Sidecar`s.
+
+## Proposal
+
+Augment `TaskRunSpec` and `PipelineTaskRunSpec` with a mapping of `Step` names to overrides
+and a mapping of `Sidecar` names to overrides.
+
+## Design Details
+
+```go
+import corev1 "k8s.io/api/core/v1"
+
+type TaskRunSpec struct {
+   ...
+   // A map of Step names to ResourceRequirements for that Step.
+   // A single entry with key "*" will apply the ResourceRequirements to all Steps.
+   // If ResourceRequirements are specified in both a Step and StepResources,
+   // the value from the StepResources will be used.
+   StepResources map[string]corev1.ResourceRequirements
+
+   // A map of Sidecar names to ResourceRequirements to apply to that Sidecar.
+   // A single entry with key "*" will apply the ResourceRequirements to all Sidecars.
+   // If ResourceRequirements are specified in both a Sidecar and SidecarResources,
+   // the value from the SidecarResources will be used.
+   SidecarResources map[string]corev1.ResourceRequirements
+}
+
+type PipelineTaskRunSpec struct {
+  ...
+   // A map of Step names to ResourceRequirements for that Step.
+   // A single entry with key "*" will apply the ResourceRequirements to all Steps.
+   // If ResourceRequirements are specified in both a Step and StepResources,
+   // the value from the StepResources will be used.
+   StepResources map[string]corev1.ResourceRequirements
+
+   // A map of Sidecar names to ResourceRequirements to apply to that Sidecar.
+   // A single entry with key "*" will apply the ResourceRequirements to all Sidecars.
+   // If ResourceRequirements are specified in both a Sidecar and SidecarResources,
+   // the value from the SidecarResources will be used.
+   SidecarResources map[string]corev1.ResourceRequirements
+}
+```
+
+### Example Task and TaskRun
+
+Example `Task`:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Task
+metadata:
+  name: image-build-task
+spec:
+  steps:
+    - name: build
+      image: gcr.io/kaniko-project/executor:latest
+      command:
+        - /kaniko/executor
+```
+
+Example `TaskRun`:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: TaskRun
+metadata:
+  name: image-build-taskrun
+spec:
+  taskRef:
+    name: image-build-task
+  stepResources:
+    build:
+      resources:
```

Comment:
nit: would there be a top level `resources` field with the syntax suggested above or would this be:

```yaml
  stepResources:
    build:
      requests:
        memory: 1Gi
```

Reply:
I would prefer to stick with the top level `resources` field now that I've updated the proposal to use a new struct `TaskRunStepOverride`.